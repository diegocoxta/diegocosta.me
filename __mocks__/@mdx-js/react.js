module.exports = {
  MDXProvider: jest.fn(({ children }) => children),
};
