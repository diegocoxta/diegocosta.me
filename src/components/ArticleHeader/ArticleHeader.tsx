import React from 'react';
import { Link } from 'gatsby';
import styled from 'styled-components';
import kebabCase from 'lodash.kebabcase';

import translations from './translations';

const Container = styled.header`
  margin: 20px 0 0 0;
`;

const Title = styled.h3`
  font-family: 'Raleway', sans-serif;
  font-weight: 700;
  font-size: 36px;
  margin: 0;
`;

const CustomLink = styled(Link)`
  color: ${({ theme }) => theme.titleColor};
  box-shadow: none;
  text-decoration: none;

  :hover,
  :focus {
    border-bottom: 1px solid ${({ theme }) => theme.titleColor};
    outline: none;
  }
`;

const TagLink = styled(Link)`
  text-decoration: none;
  box-shadow: none;
  font-size: 14px;
  color: ${({ theme }) => theme.accentColor};
  font-weight: 700;
  text-transform: lowercase;

  :hover,
  :focus {
    border-bottom: 1px solid ${({ theme }) => theme.accentColor};
    outline: none;
  }
`;

const TagList = styled.ul`
  display: flex;
  flex-wrap: wrap;
  margin-bottom: 0;
  padding: 0;
`;

const TagItem = styled.li`
  list-style: none;
  padding: 0 10px 0 0;
  margin-bottom: 5px;
`;

const Details = styled.p`
  color: ${({ theme }) => theme.subtitleColor};
  font-size: 18px;
  margin: 5px 0;
`;

export interface ArticleHeaderProps {
  url?: string | null;
  title: string;
  readingTime: number;
  lang?: string | null;
  date: string;
  tags?: string[] | null;
}

export default function ArticleHeader(props: ArticleHeaderProps): React.ReactElement {
  const texts = translations[props.lang ?? 'en'];
  const readingTime = props.readingTime < 1 ? texts.lessThan1Minute : `${props.readingTime.toFixed()} ${texts.minutes}`;

  const Tags = props.tags && (
    <TagList data-testid="article-header-tags">
      {props.tags.map((tag: string, index: number) => (
        <TagItem key={`${index}-${tag}`} data-testid="article-header-tag">
          <TagLink to={`/tags/${kebabCase(tag ?? '')}`}>#{kebabCase(tag ?? '')}</TagLink>
        </TagItem>
      ))}
    </TagList>
  );

  const linkWithLanguage = (url: string) => `/${props.lang ? props.lang : ''}${url}`;

  return (
    <Container>
      <Title>
        {props.url ? (
          <CustomLink to={linkWithLanguage(props.url)} data-testid="article-header-custom-link">
            {props.title}
          </CustomLink>
        ) : (
          props.title
        )}
      </Title>
      <Details>
        <span>
          {props.date} Â· {readingTime} {texts.ofReading}
        </span>
      </Details>
      {Tags}
    </Container>
  );
}
