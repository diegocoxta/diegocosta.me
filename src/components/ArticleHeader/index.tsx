import React from 'react';
import { useStaticQuery, graphql } from 'gatsby';

import ArticleHeader, { ArticleHeaderProps } from './ArticleHeader';

export default (props: ArticleHeaderProps): React.ReactElement => {
  const {
    site: {
      siteMetadata: { language },
    },
  } = useStaticQuery(
    graphql`
      query ArticleHeaderComponent {
        site {
          siteMetadata {
            language
          }
        }
      }
    `
  );

  return <ArticleHeader {...props} lang={props.lang || language} />;
};
