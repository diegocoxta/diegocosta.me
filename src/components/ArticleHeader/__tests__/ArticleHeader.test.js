import React from 'react';
import { render } from '@testing-library/react';

import ArticleHeader from '../ArticleHeader';

describe('<ArticleHeader />', () => {
  it('should render properly', () => {
    const { baseElement, getByTestId, getByText, getAllByTestId } = render(
      <ArticleHeader
        title="Awesome Article"
        url="/awesome-article"
        readingTime={5}
        lang="en"
        date="20/07/2020"
        tags={['jest', 'testing-library']}
      />
    );
    expect(getByText('Awesome Article')).toBeTruthy();
    expect(getByText('20/07/2020 · 5 minutes of reading')).toBeTruthy();
    expect(getByTestId('article-header-custom-link').href).toBe('http://localhost/en/awesome-article');
    expect(getByTestId('article-header-tags')).toBeTruthy();
    expect(getAllByTestId('article-header-tag').length).toEqual(2);
    expect(baseElement).toMatchSnapshot();
  });

  it('should not render the tags', () => {
    const { baseElement, queryByTestId, queryAllByTestId, getByText } = render(
      <ArticleHeader title="Awesome Article" url="/awesome-article" readingTime={5} lang="pt" date="20/07/2020" />
    );

    expect(queryByTestId('article-header-tags')).toBeFalsy();
    expect(getByText('20/07/2020 · 5 minutos de leitura')).toBeTruthy();
    expect(queryAllByTestId('article-header-tag').length).toEqual(0);
    expect(baseElement).toMatchSnapshot();
  });

  it('should not render the link', () => {
    const { baseElement, queryByTestId } = render(
      <ArticleHeader title="Awesome Article" readingTime={5} lang="en" date="20/07/2020" />
    );

    expect(queryByTestId('article-header-custom-link')).toBeFalsy();
    expect(baseElement).toMatchSnapshot();
  });

  it('should render properly in Poruguese', () => {
    const { baseElement, getByText } = render(
      <ArticleHeader title="Awesome Article" url="/awesome-article" readingTime={5} lang="pt" date="20/07/2020" />
    );

    expect(getByText('20/07/2020 · 5 minutos de leitura')).toBeTruthy();
    expect(baseElement).toMatchSnapshot();
  });

  it('should render the language link properly', () => {
    const { getByTestId } = render(
      <ArticleHeader
        title="Awesome Article"
        url="/awesome-article"
        readingTime={5}
        lang="pt"
        date="20/07/2020"
        tags={['jest', 'testing-library']}
      />
    );

    expect(getByTestId('article-header-custom-link').href).toBe('http://localhost/pt/awesome-article');
  });
});
