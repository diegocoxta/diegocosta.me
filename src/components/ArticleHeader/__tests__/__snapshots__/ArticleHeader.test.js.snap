// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleHeader /> should not render the link 1`] = `
<body>
  .c0 {
  margin: 20px 0 0 0;
}

.c1 {
  font-family: 'Raleway',sans-serif;
  font-weight: 700;
  font-size: 36px;
  margin: 0;
}

.c2 {
  font-size: 18px;
  margin: 5px 0;
}

<div>
    <header
      class="c0"
    >
      <h3
        class="c1"
      >
        Awesome Article
      </h3>
      <p
        class="c2"
      >
        <span>
          20/07/2020
           路 
          5 minutes
           
          of reading
        </span>
      </p>
    </header>
  </div>
</body>
`;

exports[`<ArticleHeader /> should not render the tags 1`] = `
<body>
  .c0 {
  margin: 20px 0 0 0;
}

.c1 {
  font-family: 'Raleway',sans-serif;
  font-weight: 700;
  font-size: 36px;
  margin: 0;
}

.c2 {
  box-shadow: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:hover,
.c2:focus {
  border-bottom: 1px solid;
  outline: none;
}

.c3 {
  font-size: 18px;
  margin: 5px 0;
}

<div>
    <header
      class="c0"
    >
      <h3
        class="c1"
      >
        <a
          class="c2"
          data-testid="article-header-custom-link"
          href="/pt/awesome-article"
        >
          Awesome Article
        </a>
      </h3>
      <p
        class="c3"
      >
        <span>
          20/07/2020
           路 
          5 minutos
           
          de leitura
        </span>
      </p>
    </header>
  </div>
</body>
`;

exports[`<ArticleHeader /> should render properly 1`] = `
.c0 {
  margin: 20px 0 0 0;
}

.c1 {
  font-family: 'Raleway',sans-serif;
  font-weight: 700;
  font-size: 36px;
  margin: 0;
}

.c2 {
  box-shadow: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:hover,
.c2:focus {
  border-bottom: 1px solid;
  outline: none;
}

.c6 {
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  font-size: 14px;
  font-weight: 700;
  text-transform: lowercase;
}

.c6:hover,
.c6:focus {
  border-bottom: 1px solid;
  outline: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-bottom: 0;
  padding: 0;
}

.c5 {
  list-style: none;
  padding: 0 10px 0 0;
  margin-bottom: 5px;
}

.c3 {
  font-size: 18px;
  margin: 5px 0;
}

<body>
  <div>
    <header
      class="c0"
    >
      <h3
        class="c1"
      >
        <a
          class="c2"
          data-testid="article-header-custom-link"
          href="/en/awesome-article"
        >
          Awesome Article
        </a>
      </h3>
      <p
        class="c3"
      >
        <span>
          20/07/2020
           路 
          5 minutes
           
          of reading
        </span>
      </p>
      <ul
        class="c4"
        data-testid="article-header-tags"
      >
        <li
          class="c5"
          data-testid="article-header-tag"
        >
          <a
            class="c6"
            href="/tags/jest"
          >
            #
            jest
          </a>
        </li>
        <li
          class="c5"
          data-testid="article-header-tag"
        >
          <a
            class="c6"
            href="/tags/testing-library"
          >
            #
            testing-library
          </a>
        </li>
      </ul>
    </header>
  </div>
</body>
`;

exports[`<ArticleHeader /> should render properly in Poruguese 1`] = `
<body>
  .c0 {
  margin: 20px 0 0 0;
}

.c1 {
  font-family: 'Raleway',sans-serif;
  font-weight: 700;
  font-size: 36px;
  margin: 0;
}

.c2 {
  box-shadow: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:hover,
.c2:focus {
  border-bottom: 1px solid;
  outline: none;
}

.c3 {
  font-size: 18px;
  margin: 5px 0;
}

<div>
    <header
      class="c0"
    >
      <h3
        class="c1"
      >
        <a
          class="c2"
          data-testid="article-header-custom-link"
          href="/pt/awesome-article"
        >
          Awesome Article
        </a>
      </h3>
      <p
        class="c3"
      >
        <span>
          20/07/2020
           路 
          5 minutos
           
          de leitura
        </span>
      </p>
    </header>
  </div>
</body>
`;
